crates/wasi-nn/src/backend/winml.rs

```diff
@@ -226,3 +205,180 @@ fn convert_i64(i: i64) -> Result<u32, BackendError> {
         anyhow::anyhow!("unable to convert dimension to u32: {d}").into()
     })
 }
+
+// Convert from wasi-nn tensor to WinML tensor.
+fn to_inspectable(tensor: &Tensor, shape: IVectorView<i64>) -> Result<IInspectable, Error> {
+    match tensor.ty {
+        crate::wit::types::TensorType::Fp16 => unsafe {
+            let data = std::slice::from_raw_parts(
+                tensor.data.as_ptr() as *const f32,
+                tensor.data.len() / size_of::<f32>(),
+            );
+            TensorFloat16Bit::CreateFromArray(&shape, data)?.cast::<IInspectable>()
+        },
+        crate::wit::types::TensorType::Fp32 => unsafe {
+            let data = std::slice::from_raw_parts(
+                tensor.data.as_ptr() as *const f32,
+                tensor.data.len() / size_of::<f32>(),
+            );
+            TensorFloat::CreateFromArray(&shape, data)?.cast::<IInspectable>()
+        },
+        crate::wit::types::TensorType::I64 => unsafe {
+            let data = std::slice::from_raw_parts(
+                tensor.data.as_ptr() as *const i64,
+                tensor.data.len() / size_of::<i64>(),
+            );
+            TensorInt64Bit::CreateFromArray(&shape, data)?.cast::<IInspectable>()
+        },
+        _ => unimplemented!(),
+    }
+}
+
+// Convert from WinML tensor to wasi-nn tensor.
+fn to_tensor(inspectable: IInspectable, tensor_kind: TensorKind) -> Result<Tensor, BackendError> {
+    let tensor = match tensor_kind {
+        TensorKind::Float16 => {
+            let output_tensor = inspectable.cast::<TensorFloat16Bit>()?;
+            let dimensions = dimensions_as_u32(&output_tensor.Shape()?)?;
+            let view = output_tensor.GetAsVectorView()?;
+            // TODO: Move to f16 when it's available in stable.
+            let mut data = Vec::with_capacity(view.Size()? as usize * size_of::<f32>());
+            for f in view.into_iter() {
+                data.extend(f.to_le_bytes());
+            }
+            Tensor {
+                ty: TensorType::Fp16,
+                dimensions,
+                data,
+            }
+        }
+        TensorKind::Float => {
+            let output_tensor = inspectable.cast::<TensorFloat>()?;
+            let dimensions = dimensions_as_u32(&output_tensor.Shape()?)?;
+            let view = output_tensor.GetAsVectorView()?;
+            let mut data = Vec::with_capacity(view.Size()? as usize * size_of::<f32>());
+            for f in view.into_iter() {
+                data.extend(f.to_le_bytes());
+            }
```

Comment:
Not sure if it's any faster, but this would avoid the size calculation:

```rust
let data = view.into_iter().map(f32::to_le_bytes).collect();
```

We might want to note somewhere that the LE ordering is here to match WebAssembly's ordering, not the platform we're running on (in which case always doing LE could be incorrect).

Reply:
`view`'s type is `IVectorView`, not a rust collection, so the short version doesn't work here.

Reply:
`flat_map` works. Thanks.